Make cgroups more robust and provide the override similar to ES

guyboertje opened the pull request:
and Kibana

This should go out as soon as possible and backport to 5.6

yaauie commented on logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb:
```diff
@@ -6,130 +6,234 @@
+         :cpuacct => {},
+         :cpu => {}
+        }
+        cpuacct_controller.fill(cgroups_stats[:cpuacct])
```
I would prefer inverting the `#fill` method (which mutates a passed hash) into a `#to_hash` method (that takes no argument and returns a hash), especially since it is only used here, which could easily be refactored to be the following without creating any temporary objects:

~~~ ruby
cgroups_stats = {
  :cpuacct => cpuacct_controller.to_hash,
  :cpu     => cpu_controller.to_hash,
}

~~~

I believe this would also simplify our unit tests, which currently have to juggle a bit to validate side-effects  (e.g. providing empty hash to inject data into).

yaauie commented on logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb:
```diff
@@ -6,130 +6,234 @@
-
-      def cgroup_cpu_fs_quota_micros(control_group)
-        read_first_line(::File.join(PROC_CGROUP_CPUACCT_DIR, control_group,  CPU_FS_QUOTA_US_FILE)).to_i
+      def exist?(path)
```
that nearly every method in this module takes a `path` argument, and the path is a result of combining instance variables with something in an implementation's other method(s), leads me to believe we haven't found the most useful level of abstraction.

yaauie commented on logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb:
```diff
@@ -6,130 +6,234 @@
+      CONTROL_GROUP_RE = Regexp.compile("\\d+:([^:,]+(?:,[^:,]+)?):(/.*)")
+      CONTROLLER_SEPARATOR_RE = ","
+
+      def initialize
```
All of the statically-defined `Pathname`s could be constants here, since `Pathname` is immutable.

yaauie commented on logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb:
```diff
@@ -6,130 +6,234 @@
+          yield
+        else
+          message = "File #{path} cannot be found, "
+          if override.nil?
```
not consistent/compatible with `NullOverride` abstraction we use when creating `UnhandledResource`, which includes this module. From my understanding, we don't actually _use_ `UnhandledResource#check_path` anywhere currently, but this could be problematic if we do.

guyboertje replied:
We know from the code that only deal with `cpu` and `cpuacct` and this code has not changed in two years so I think the chances of expansion to include values from other controllers is less likely. However, if that change was required then its obvious that one would create a new controller and remove the UnhandledResource for that controller key.

guyboertje commented on logstash-core/lib/logstash/instrument/periodic_poller/cgroup.rb:
```diff
@@ -6,130 +6,234 @@
           # multiples controls, same hierarchy
           controllers = matches[1].split(CONTROLLER_SEPARATOR_RE)
-          controllers.each_with_object(response) { |controller| response[controller] = matches[2] }
+          controllers.each_with_object(response) do |controller|
```
I agree, that did look weird.

guyboertje commented:
@yaauie 
After reading your wise words I had another look.

- I removed Pathname as it was not giving much in return.
- I changed the stubbing in the tests to work on IO and File instead of the methods
- I re-organised the CpuResource and friends and divorced the UnhandledResource from them.
- I made CpuStat do its own update and gave it a too hash method.
- I changed the groups handling to not be reliant (and fragile) on it having `cpu` and `cpuacct` keys.

guyboertje commented:
FYI I overwrote the cgroup.rb contents in a previously failing google compute ubuntu instance - works in practice.

yaauie commented:
LGTM
